WordPress/Classes/ViewRelated/Notifications/Tools/NotificationMediaDownloader.swift

```diff
@@ -1,55 +1,110 @@
 import Foundation
 
 
+/**
+*  @class           NotificationMediaDownloader
+*  @brief           The purpose of this class is to provide a simple API to download assets from the web.
+*  @details         Assets are downloaded, and resized to fit a maximumWidth, specified in the initial download call.
+*                   Internally, images get downloaded and resized: both copies of the image get cached.
+*                   Since the user may rotate the device, we also provide a second helper (resizeMediaWithIncorrectSize),
+*                   which will take care of resizing the original image, to fit the new orientation.
+*/
+
 @objc public class NotificationMediaDownloader : NSObject
 {
+    //
+    // MARK: - Public Methods
+    //
     deinit {
         downloadQueue.cancelAllOperations()
     }
     
-    public init(maximumImageWidth: CGFloat) {
-        downloadQueue       = NSOperationQueue()
-        resizeQueue         = dispatch_queue_create("org.wordpress.notifications.media-downloader", DISPATCH_QUEUE_CONCURRENT)
-        mediaMap            = [NSURL: UIImage]()
-        retryMap            = [NSURL: Int]()
-        maxImageWidth       = maximumImageWidth
-        responseSerializer  = AFImageResponseSerializer() as AFImageResponseSerializer
-        super.init()
+    /**
+    *  @brief       Downloads a set of assets, resizes them (if needed), and hits a completion block.
+    *  @details     The completion block will get called just once all of the assets are downloaded, and properly sized.
+    *
+    *  @param       urls            Is the collection of unique Image URL's we'd need to download.
+    *  @param       maximumWidth    Represents the maximum width that a returned image should have
+    *  @param       completion      Is a closure that will get executed once all of the assets are ready
+    */
+    public func downloadMedia(#urls: Set<NSURL>, maximumWidth: CGFloat, completion: SuccessBlock) {
+        let missingUrls         = filter(urls) { self.shouldDownloadImage(url: $0) }
+        let group               = dispatch_group_create()
+        var shouldHitCompletion = !missingUrls.isEmpty
+
+        for url in missingUrls {
+            
+            dispatch_group_enter(group)
+            
+            downloadImage(url) {
+                self.originalImagesMap[url] = $0
```

Comment:
It looks like `downloadImage` only calls its block on success and if an image fails then `dispach_group_leave(group)` is never called.